File: openslides/websocket.py
"""
Module for handeling the websocket connects.

Each client is an instance of the Consumer class.
"""
from __future__ import annotations

import asyncio
import json
from typing import Any, Dict, List, Set

import websockets
from collections import defaultdict

from .actions import ActionData, ValidationError, handle_actions, prepare_actions
from .db import get_database
from .utils import debug


class Client:
    all_clients: Set["Client"] = set()

    def __init__(self, websocket: websockets.WebSocketServerProtocol) -> None:
        self.websocket = websocket

    def __enter__(self) -> "Client":
        self.all_clients.add(self)
        return self

    def __exit__(self, *args: Any) -> None:
        self.all_clients.remove(self)

    async def connected(self) -> None:
        """
        Called when the websocket is opened.
        """
        all_data: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
        for collection, data in (await get_database()).items():
            for element in data.values():
                all_data[collection].append(element)

        await self.send(
            {"type": "autoupdate", "changed": all_data, "deleted": {}, "all_data": True}
        )

    async def recv(self, message: Dict[str, Any]) -> None:
        """
        Our demo handels all incomming data as action. It has to have the format

        {
            "id": "message_id",
            "actions": [
                {
                    "action": "ACTION_NAME",
                    "payload": {
                        "SOME": "DATA"
                    }
                }
            ]
        }
        """
        message_id = message["id"]
        action_data = await prepare_actions(message["actions"])
        try:
            await handle_actions(action_data)
        except ValidationError as err:
            await self.send({"type": "response", "error": str(err), "id": message_id})
        else:
            await self.send({"type": "response", "id": message_id})

    async def send(self, message: Dict[str, Any]) -> None:
        """
        Sends data to the client.
        """
        await self.websocket.send(json.dumps(message))

    @classmethod
    async def send_to_all(cls, message: Dict[str, Any]) -> None:
        """
        Sends data to all connected clients.
        """
        encoded = json.dumps(message)
        for client in cls.all_clients:
            await client.websocket.send(encoded)


async def handler(websocket: websockets.WebSocketServerProtocol, path: str) -> None:
    client = Client(websocket).__enter__()
    try:
        debug(f"New connection, currently {len(Client.all_clients)} connected clients")
        await client.connected()
        async for message in websocket:
            await client.recv(json.loads(message))
    except websockets.exceptions.ConnectionClosed:
        pass
    finally:
        client.__exit__()
        debug(f"Lost connection, currently {len(Client.all_clients)} connected clients")


def serve(host: str, port: int) -> None:
    start_server = websockets.serve(handler, host, port)
    asyncio.get_event_loop().run_until_complete(start_server)
    print(f"Started Server on {host}:{port}")
    asyncio.get_event_loop().run_forever()

File: go/websocket.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

var clients map[*Client]struct{}
var clientsLock sync.Mutex

func init() {
	clients = make(map[*Client]struct{})
}

// AutoupdateMessage is a message to the client when the server sends elements.
type AutoupdateMessage struct {
	Type    string                   `json:"type"`
	Changed map[string][]interface{} `json:"changed"`
	Deleted map[string][]int         `json:"deleted"`
	AllData bool                     `json:"all_data"`
}

// ErrorMessage is a message send to the client when there is an error.
type ErrorMessage struct {
	Type       string `json:"type"`
	ResponseID string `json:"response-id"`
	Error      string `json:"error"`
}

// ResponseActionMessage is a message to the client when the client sends some actions.
type ResponseActionMessage struct {
	Type       string                   `json:"type"`
	ResponseID string                   `json:"response-id"`
	Responses  []map[string]interface{} `json:"responses"`
}

// RecvMessage contains a message id and a list of ActionData.
type RecvMessage struct {
	ID      string       `json:"id"`
	Actions []ActionData `json:"actions"`
}

// Client represents a connection from one client.
type Client struct {
	conn     *websocket.Conn
	sendChan chan []byte
}

// NewClient creates a new client connection
func NewClient(conn *websocket.Conn) *Client {
	client := Client{conn, make(chan []byte)}
	go client.handleSend()
	return &client
}

func (c *Client) register() {
	clientsLock.Lock()
	clients[c] = struct{}{}
	clientsLock.Unlock()
	debug(
		fmt.Sprintf("New connection, currently %d connected clients", len(clients)),
	)
}

func (c *Client) unregister() {
	clientsLock.Lock()
	delete(clients, c)
	clientsLock.Unlock()
	debug(
		fmt.Sprintf("Lost connection, currently %d connected clients", len(clients)),
	)
}

func (c *Client) connected() {
	// client connects to server
	allData := make(map[string][]interface{})
	for collection, data := range getDatabase() {
		allData[collection] = make([]interface{}, 0)
		for _, element := range data {
			allData[collection] = append(allData[collection], element)
		}
	}
	c.send(AutoupdateMessage{
		Type:    "autoupdate",
		Changed: allData,
		Deleted: make(map[string][]int),
		AllData: true,
	})
}

func (c *Client) send(message interface{}) error {
	b, err := json.Marshal(message)
	if err != nil {
		return err
	}
	c.sendChan <- b
	return nil
}

func (c *Client) recv(message RecvMessage) {
	prepareActions(message.Actions, time.Now().Unix())
	allData, returnValues, err := handleActions(message.Actions)
	if err != nil {
		c.send(ErrorMessage{
			Type:       "response",
			Error:      err.Error(),
			ResponseID: message.ID,
		})
		return
	}
	c.send(ResponseActionMessage{
		Type:       "response",
		ResponseID: message.ID,
		Responses:  returnValues,
	})
	InformChangedElements(allData)
}

func (c *Client) handleSend() {
	for message := range c.sendChan {
		c.conn.WriteMessage(websocket.TextMessage, message)
	}
}

// SendToAll sends a message to all clients
func SendToAll(message interface{}) error {
	b, err := json.Marshal(message)
	if err != nil {
		return err
	}
	for client := range clients {
		client.sendChan <- b
	}
	return nil
}

func wsHandler(w http.ResponseWriter, r *http.Request) {
	conn, err := websocket.Upgrade(w, r, w.Header(), 1024, 1024)
	if err != nil {
		http.Error(w, "Could not open websocket connection", http.StatusBadRequest)
	}

	client := NewClient(conn)
	defer close(client.sendChan)
	defer conn.Close()
	client.register()
	defer client.unregister()

	client.connected()
	for {
		messageType, p, err := conn.ReadMessage()
		if err != nil {
			if _, ok := err.(*websocket.CloseError); ok {
				break
			}
			log.Panicf("Websocket error: %s\n", err)
			return
		}
		if messageType != websocket.TextMessage {
			log.Panicf("Supports only text messages\n")
		}
		var message RecvMessage
		err = json.Unmarshal(p, &message)
		if err != nil {
			log.Panicf("Can not convert message to json: %s\n", err)
		}
		client.recv(message)
	}

}

// Serve starts the webserver and listens for websocket connections
func Serve(host string) {
	http.HandleFunc("/", wsHandler)

	fmt.Printf("Started Server on %s\n", host)
	panic(http.ListenAndServe(host, nil))
}
